### PR TITLE
Make cucumber node maps easier to read by updating the syntax

### PR DESCRIPTION
**Pull request by emiltin**
# Issue

Make Cucumber node maps easier to read by changing the syntax. #2925.

## Tasklist
 - [x] update support code
 - [x] update all cucumber feature files
 - [x] review
 - [x] adjust for for comments

## Requirements / Relations
 Don't think there are any.

### CODE REVIEW

**Comment by emiltin** [comment]
The new syntax allows the addition of eg. dots to indicate the visual layout of ways. These have not been added yet.

For reference, node maps where changed using the following Ruby script:

```ruby
def replace_file file
  puts file
  lines = File.readlines(file)
  
  table = false
  first = false
  indent = nil
  
  lines.each_with_index do |line, i|
    if table
      m = line.match /^(\s+)(\|.+\|)/     # still in table content?
      if m
        indent = m[1]
        content = m[2].scan( /[^\|]+/ ).map do |t|    # find content between pipes
          t = t.strip[0] || ' '                       # strip whitespace and use " " for empty cells
        end.join ' '                                  # join with a pace between neighbouring cells
        
        content.rstrip!
        
        changed_line = "#{indent}#{content}".rstrip
        
        if first
          # if start of table content, add doc string initiator
          str = %!#{indent}"""\n#{changed_line}\n!
        else
          str = %!#{changed_line}\n!
        end
        lines[i] = str
        first = false
      else
        # not in table content anymore, add doc string terminator
        lines[i] = %!#{indent}"""\n#{lines[i]}! unless first
        table = false
      end
    else
      if line =~ /the node map/     # start of a node map table?
        table = true
        first = true
      end
    end
  end
  
  File.open(file, 'w') { |f| f.write(lines.join) }
end


files = Dir.glob("./features/**/*.feature")
files.each do |file|
  puts file
  replace_file file
end
```
 
A few tests has comment lines using '#', these were removed manually first.

A few tests used a peculiar syntax where the node ids was set in the node map using syntax like "a:1", "b:2", etc. This has been change so node ids is set in a separate table.

**Comment by emiltin** [comment]
I get a single failing test locally, but I don't understand what the test is doing.

```
1) Scenario: osrm-contract - Passing base file - features/options/contract/datasources.feature:27
   Step: Then datasource names should contain "lua profile,25_osrmcontract_passing_base_file_speeds" - features/options/contract/datasources.feature:29
   Step Definition: features/step_definitions/options.js:93
   Message:
     AssertionError: 'lua profile,27_osrmcontract_passing_base_file_speeds' == 'lua profile,25_osrmcontract_passing_base_file_speeds'
         at World.<anonymous> (/Users/emil/code/osrm-backend/features/step_definitions/options.js:95:16)
         at _combinedTickCallback (node.js:370:9)
         at process._tickCallback (node.js:401:11)
```

**Comment by emiltin** [comment]
@MoKob anything additional you want me to do here?

**Comment by MoKob** [comment]
At least we need to have the test-setup working again. Seems to be a minor linting error (unused variable).
If that is taken care of, we can do the full review process.

**Comment by emiltin** [comment]
Travis chews on commit b8226d1, even though I committed dde3fc3 above - and it seem failing tests refer to versions that are not in my local repo anymore. Could it be because I did a forced push previously?

**Comment by TheMarex** [comment]
@emiltin I would recommend rebasing and master and then force pushing again. It seems travis has had a hiccup.

**Comment by emiltin** [comment]
Test setup is now working, all tests passing except the following one related to speed files:

```gherkin
    Background:
        Given the profile "testbot"
        Given the extract extra arguments "--generate-edge-lookup"
        And the node map
            """
            a b
            """
        And the ways
            | nodes |
            | ab    |
        And the speed file
        """
        1,2,27
        2,1,27
        2,3,27
        3,2,27
        1,4,27
        4,1,27
        """
        And the data has been extracted

    Scenario: osrm-contract - Passing base file
        When I run "osrm-contract --segment-speed-file {speeds_file} {processed_file}"
        Then datasource names should contain "lua profile,25_osrmcontract_passing_base_file_speeds"
        And it should exit successfully
```

I don't understand what is the following line is suppose to verify?

```gherkin
        Then datasource names should contain "lua profile,25_osrmcontract_passing_base_file_speeds"
```

The cucumber error is:
```
1) Scenario: osrm-contract - Passing base file - features/options/contract/datasources.feature:27
   Step: Then datasource names should contain "lua profile,25_osrmcontract_passing_base_file_speeds" - features/options/contract/datasources.feature:29
   Step Definition: features/step_definitions/options.js:93
   Message:
     AssertionError: 'lua profile,27_osrmcontract_passing_base_file_speeds' == 'lua profile,25_osrmcontract_passing_base_file_speeds'
         at World.<anonymous> (/Users/emil/code/osrm-backend/features/step_definitions/options.js:95:16)
         at _combinedTickCallback (node.js:370:9)
         at process._tickCallback (node.js:401:11)
```

The speed data contains the number 27 and the speed file on disk contains '27', but the test seem to look for the string '25'.

**Comment by emiltin** [comment]
Do you want commits squashed?

**Comment by MoKob** [comment]
I am always in favour of squashing commits.

``` Then datasource names should contain "lua profile,25_osrmcontract_passing_base_file_speeds"```
is testing for edge-lookup files. When we extract, `--generete-edge-lookup` provides a file that can be used to apply penalties/updates to the graph from other sources.

I can't see why the tests breaks, though.
Did something change the ID's of nodes generated? That would be my first guess.

File: features/guidance/anticipate-lanes.feature [hunk 1]
```diff
@@ -565,43 +609,45 @@ Feature: Turn Lane Guidance
...
-            |   |   |   |   |   |
-            |   |   |   |   |   |
-            |   |   |   |   |   |
-            |   |   |   |   |   |
-            |   |   |   |   |   |
-            |   |   | c |   |   |
+            """
+                a
+            x b   d y
+
```
Comment (MoKob) [inline features/guidance/anticipate-lanes.feature]: this would be a good candidate to show the usage of `.` to indicate roads. Here it is really hard to follow without the visual aid, in addition it would be good to have it in to see it working.

File: features/step_definitions/data.js [hunk 2]
```diff
@@ -48,41 +48,26 @@ module.exports = function () {
...
+            cb();
         };
 
-        table.raw().forEach((row, ri) => {
-            row.forEach((name, ci) => {
-                q.defer(addNode, name, ri, ci);
+        docstring.split(/\n/).forEach( (row,ri) => {
+            row.split('').forEach( (cell,ci) => {
+                if( cell.match(/[a-z0-9]/) ) {
+                    q.defer(addNode, cell, ri, ci*0.5);
```
Comment (MoKob) [inline features/step_definitions/data.js]: This 0.5 indicates that we need a space between two nodes?

**Comment by emiltin** [comment]
Why is test is looking for the string "25_osrmcontract_passing_base_file_speeds" in the speed file? Where does the 25 come from? It doesn't seem to relate to IDs, nor to the speeds specified in the file.

**Comment by MoKob** [comment]
The name is for the caching infrastructure. It describes the test you are looking at.

In your case it starts on line 25 of `features/options/contract/datasources.feature` and is named `osrm-contract - Passing base file`. This is distilled into the `25_osrmcontract_passing_base_file_speeds` you see.

**Comment by emiltin** [comment]
I see. It seems a a bit brittle to me to have the test include what line number it's starting on. But for now I'm just going to update the line number.

File: features/step_definitions/data.js [hunk 3]
```diff
@@ -48,41 +48,26 @@ module.exports = function () {
...
-
-                cb();
+    this.Given(/^the node map$/, (docstring, callback) => {
+        var q = d3.queue();
+
+        var addNode = (name, ri, ci, cb) => {
+            var lonLat = this.tableCoordToLonLat(ci, ri);
+            if (name.match(/[a-z]/) ) {
+                if (this.nameNodeHash[name]) throw new Error(util.format('*** duplicate node %s', name));
+                this.addOSMNode(name, lonLat[0], lonLat[1], null);
```
Comment (MoKob) [inline features/step_definitions/data.js]: Is this removing the `node:id` feature here, or is this the combination with https://github.com/Project-OSRM/osrm-backend/pull/2979/files#diff-c41ecde0b11b4c1dc39c3b4a6f814478R109? Having IDs is kind of important, potentially. But I don't see that we need to do it in the docstring. 
If this is what happens in added table, I would be totally fine with it. No need to have it in the docstring.

Comment (emiltin) [inline features/step_definitions/data.js]: Previously, id could be set in the node map using the special syntax 'a:14'. This doesn't work with doc strings, since you cannot insert extra characters.
I changed it so you can specify IDs using the existing 'And the nodes' tables, using an 'id' column. There are a few tests that uses this.

File: features/step_definitions/data.js (hunk 2 above)
Comment (emiltin) [inline features/step_definitions/data.js]: Yes. This is to ensure that the map doesn't look distorted (this was part of the original intention behind the PR). eg. that a square will look like a square and not a rectangle. Compare:

```gherkin
And the node map
    """
    a . b
    .   .
    c . d
    """

And the node map
    """
    a.b
    . .
    c.d
    """
```

With the current implementation, you're allowed to place nodes anywhere, so horizontally you can place nodes with 0.5 times the grid distance, vertically only on the grid. The following is valid and will result in distance of ac=1, cd=1, but ab=0.5:

```gherkin
And the node map
    """
    ab
    c d
    """
```

An alternative would be to enforce placement only on the grid, ie. always have spacing between nodes horizontally.

File: features/guidance/anticipate-lanes.feature (hunk 1 above)
Comment (emiltin) [inline features/guidance/anticipate-lanes.feature]: Right. My approach was to not add these visual clues in the first commit, since adding them too all tests is a really big job. But I'm fine with adding them to a select set of scenarios.

Comment (MoKob) [inline features/guidance/anticipate-lanes.feature]: I also don't think that they are necessary everywhere. But they should be in at least a few to make sure all works as expected.

**Comment by MoKob** [comment]
I guess ```1) Scenario: Obvious Index wigh very narrow turn to the right - features/guidance/turn.feature:1180
   Step: Given the node map - features/guidance/turn.feature:1181
   Step Definition: features/step_definitions/data.js:51
   Message:
     TypeError: docstring.split is not a function
         at World.<anonymous> (/home/travis/build/Project-OSRM/osrm-backend/features/step_definitions/data.js:66:19)
         at nextTickCallbackWith0Args (node.js:420:9)
         at process._tickCallback (node.js:349:13)
2) Scenario: Obvious Index wigh very narrow turn to the right - features/guidance/turn.feature:1197
   Step: Given the node map - features/guidance/turn.feature:1198
   Step Definition: features/step_definitions/data.js:51
   Message:
     TypeError: docstring.split is not a function
         at World.<anonymous> (/home/travis/build/Project-OSRM/osrm-backend/features/step_definitions/data.js:66:19)
         at nextTickCallbackWith0Args (node.js:420:9)
         at process._tickCallback (node.js:349:13)
3) Scenario: Obvious Index wigh very narrow turn to the left - features/guidance/turn.feature:1215
   Step: Given the node map - features/guidance/turn.feature:1216
   Step Definition: features/step_definitions/data.js:51
   Message:
     TypeError: docstring.split is not a function
         at World.<anonymous> (/home/travis/build/Project-OSRM/osrm-backend/features/step_definitions/data.js:66:19)
         at nextTickCallbackWith0Args (node.js:420:9)
         at process._tickCallback (node.js:349:13)
4) Scenario: Obvious Index wigh very narrow turn to the left - features/guidance/turn.feature:1232
   Step: Given the node map - features/guidance/turn.feature:1233
   Step Definition: features/step_definitions/data.js:51
   Message:
     TypeError: docstring.split is not a function
         at World.<anonymous> (/home/travis/build/Project-OSRM/osrm-backend/features/step_definitions/data.js:66:19)
         at nextTickCallbackWith0Args (node.js:420:9)
         at process._tickCallback (node.js:349:13)```

is due to a few testcases that have not beed adapted yet?

**Comment by emiltin** [comment]
Yes, new scenarios that we're pulled in during latest rebase.

**Comment by emiltin** [comment]
I updated docs/testing.md to reflect the new syntax. I guess I should update the wiki page  https://github.com/Project-OSRM/osrm-backend/wiki/Cucumber-Test-Suite as well? Or do you want to move that to the repo?

**Comment by MoKob** [comment]
No, https://github.com/Project-OSRM/osrm-backend/wiki/Cucumber-Test-Suite will remain. The testing part is mostly a howto for writing better cucumber tests (since I've seen quite a collection of brittle tests, failing due to unrelated changes - just like your line number experience). 

When we merge this, we need to update the guide on the website as well.

For the rest, with the changes it looks fine to me. /cc @TheMarex @daniel-j-h any further input?

**Comment by daniel-j-h** [comment]
Looks good to me! Thanks a lot @emiltin, great work! :+1:

**Comment by TheMarex** [comment]
Yup looks good to me as well. :+1: Much more readable.

**Comment by emiltin** [comment]
I'm adding visual hints to the scenarios in anticipate-lanes.feature. For some scenarios it makes sense to to use dashes and pipes:

```gherkin
        Given the node map
            """
            a
              \
            x – b – c y
                    |
                    d
            """
```

But in scenarios with ways with more unusual angles, it's better to use other symbols, eg. various dots:

```gherkin
        Given the node map
            """
            i               a
              ' .       . ' 
            j . . c . b . . x
            """
```
Each style has it's drawbacks and benefits. In light of this, do we want to stick to a single style or use what's best in each scenario?

**Comment by emiltin** [comment]
I've added visual hints to anticipate-lanes.feature: https://github.com/Project-OSRM/osrm-backend/blob/dd512812583e2105c80721d06c1a05528284a206/features/guidance/anticipate-lanes.feature.  Think it's ready to merge; once that happens we can update the wiki.
